File: Javascript/Linked List/206.反转链表.js
/*
 * @lc app=leetcode.cn id=206 lang=javascript
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    let reverse = function(pre, cur){
        if(!cur) return pre;
        // 保存 next 节点
        let next = cur.next;
        cur.next = pre;
        return reverse(cur, next);
      }
      return reverse(null, head);
};
// @lc code=end

File: Go/LinkedList/206-反转链表.go
package LinkedList

// 三指针交换，一个记录当前，一个记录前一个节点，一个记录后一个节点
func reverseList(head *ListNode) *ListNode {
	// 1 -> 2 -> 3 -> 4 -> 5
	if head == nil {
		return nil
	}
	var tmp *ListNode
	var prv *ListNode
	prv = nil
	for head != nil {
		tmp = head.Next
		head.Next = prv
		prv = head
		head = tmp
	}
	printLL(prv)
	return prv
}